src/main/java/io/stargate/sgv2/jsonapi/service/resolver/AlterTableCommandResolver.java

```diff
@@ -0,0 +1,285 @@
+          case AlterTableOperationImpl.AddVectorize av ->
+              handleAddVectorize(av, ctx.schemaObject());
+          case AlterTableOperationImpl.DropVectorize dc ->
+              handleDropVectorize(dc, ctx.schemaObject());
+          default -> throw new IllegalStateException("Unexpected value: " + operation);
+        };
+
+    OperationAttemptContainer<TableSchemaObject, SchemaAttempt<TableSchemaObject>> container =
+        new OperationAttemptContainer<>(true);
+    container.addAll(attempts);
+
+    var pageBuilder =
+        SchemaAttemptPage.<TableSchemaObject>builder()
+            .debugMode(ctx.getConfig(DebugModeConfig.class).enabled())
+            .useErrorObjectV2(ctx.getConfig(OperationsConfig.class).extendError());
+
+    return new GenericOperation<>(container, pageBuilder, new TableDriverExceptionHandler());
+  }
+
+  private List<AlterTableAttempt> handleAddColumns(
+      AlterTableOperationImpl.AddColumns ac, TableSchemaObject schemaObject) {
+    TableMetadata tableMetadata = schemaObject.tableMetadata();
+    List<AlterTableAttempt> alterTableAttempts = new ArrayList<>();
+    // check the column doesn't exists
+    ac.columns()
+        .keySet()
+        .forEach(
+            column -> {
+              if (tableMetadata
+                  .getColumn(CqlIdentifierUtil.cqlIdentifierFromUserInput(column))
+                  .isPresent()) {
+                throw SchemaException.Code.COLUMN_ALREADY_EXISTS.get(Map.of("column", column));
+              }
+            });
+
+    // New columns to be added
+    Map<String, ApiDataType> addColumns =
+        ac.columns().entrySet().stream()
+            .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().getApiDataType()));
+
+    // Vectorize config for the new columns
+    Map<String, VectorConfig.ColumnVectorDefinition.VectorizeConfig> vectorizeConfigMap =
+        ac.columns().entrySet().stream()
+            .filter(
+                e ->
+                    e.getValue() instanceof ComplexTypes.VectorType vt
+                        && vt.getVectorConfig() != null)
+            .collect(
+                Collectors.toMap(
+                    Map.Entry::getKey,
+                    e -> {
+                      ComplexTypes.VectorType vectorType = ((ComplexTypes.VectorType) e.getValue());
+                      final VectorizeConfig vectorizeConfig = vectorType.getVectorConfig();
+                      validateVectorize.validateService(vectorizeConfig, vectorType.getDimension());
+                      return new VectorConfig.ColumnVectorDefinition.VectorizeConfig(
+                          vectorizeConfig.provider(),
+                          vectorizeConfig.modelName(),
+                          vectorizeConfig.authentication(),
+                          vectorizeConfig.parameters());
+                    }));
+    final AlterTableAttempt addColumnsAttempt =
+        new AlterTableAttemptBuilder(schemaObject).addColumns(addColumns).build();
```

Comment:
It seems to me again that the Builder wouldn't be needed here; instead, 4 factory methods would suffice.
But I guess there may be some need to follow this approach? Like, incrementing `position` for later attempts?

Anyway -- what seems like code smell are all those `assert`s: we get one (and only one) alter operation, so we should have very little actual state. But yet with Builder we need to guard that unnecessary state. So there seems to be no value for Builder here.

Reply:
Using a Builder ok; will raise the issue with Aaron.